src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+        {
+            var solution = document.Project.Solution;
+            var solutionEditor = new SolutionEditor(solution);
+            var codeGenerationService = document.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var destinationSyntaxNode = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(
+                document.Project.Solution, result.Destination, default, cancellationToken).ConfigureAwait(false);
+            var (editorMap, syntaxMap) = await InitializeEditorMapsAndSyntaxMapAsync(result, solution, solutionEditor, destinationSyntaxNode, cancellationToken).ConfigureAwait(false);
+            var symbolsToPullUp = result.MemberAnalysisResults.
+                SelectAsArray(analysisResult =>
+                {
+                    if (analysisResult.Member is IPropertySymbol propertySymbol)
+                    {
+                        if (analysisResult.ChangeOriginalToPublic)
+                        {
+                            // We are pulling a non-public property, change its getter/setter to public and itself to be public.
+                            return CodeGenerationSymbolFactory.CreatePropertySymbol(
+                                propertySymbol,
+                                accessibility: Accessibility.Public,
+                                getMethod: CreatePublicGetterAndSetter(propertySymbol.GetMethod),
+                                setMethod: CreatePublicGetterAndSetter(propertySymbol.SetMethod));
+                        }
+                        else
+                        {
+                            // We are pulling a public property, filter the non-public getter/setter.
+                            return CodeGenerationSymbolFactory.CreatePropertySymbol(
+                                propertySymbol,
+                                getMethod: FilterGetterOrSetter(propertySymbol.GetMethod),
+                                setMethod: FilterGetterOrSetter(propertySymbol.SetMethod));
+                        }
+                    }
+                    else
+                    {
+                        return analysisResult.Member;
+                    }
+                });
+
+            // Add members to interface
+            var options = new CodeGenerationOptions(generateMethodBodies: false, generateMembers: false);
+            var destinationWithMembersAdded = codeGenerationService.AddMembers(destinationSyntaxNode, symbolsToPullUp, options: options, cancellationToken: cancellationToken);
+            var destinationEditor = editorMap[destinationSyntaxNode.SyntaxTree];
+            destinationEditor.ReplaceNode(destinationSyntaxNode, (syntaxNode, generator) => destinationWithMembersAdded);
+
+            // Change original members
+            foreach (var analysisResult in result.MemberAnalysisResults)
+            {
+                foreach (var syntax in syntaxMap[analysisResult.Member])
```

Comment:
nit.  consider naming this 'memberToDeclarationMap' or something more evocative of what it contains than 'syntaxMap'.